infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/namespace/KubernetesDeployments.java

```diff
@@ -582,8 +593,36 @@ private boolean happenedAfterWatcherInitialization(PodEvent event)
     containerEventsHandlers.add(handler);
   }
 
-  /** Stops watching the pods inside Kubernetes namespace. */
+  /**
+   * Start watching the logs of this deployment.
+   *
+   * @param handler is processing log messages
+   * @param podNames pods of interest for watching the logs
+   */
+  public void watchLogs(PodLogHandler handler, LogWatchTimeouts timeouts, Set<String> podNames)
+      throws InfrastructureException {
+    if (logWatcher == null) {
+      LOG.debug("start watching logs of pods '{}' of  workspace '{}'", podNames, workspaceId);
+      logWatcher =
+          new LogWatcher(clientFactory, workspaceId, namespace, podNames, executor, timeouts);
+      logWatcher.addLogHandler(handler);
+      watchEvents(logWatcher);
+    } else {
+      LOG.debug("Already watching logs of workspace [{}], just adding log handler", workspaceId);
+      logWatcher.addLogHandler(handler);
+    }
+  }
+
   public void stopWatch() {
+    stopWatch(false);
+  }
+
+  /**
+   * Stops watching the pods inside Kubernetes namespace.
+   *
+   * @param failed use true if is this stopWatch called by failed workspace. false otherwise.
```

Comment:
This is hard to understand, maybe
```suggestion
   * @param failed true if workspace startup ended in failure.
```

Perhaps also include some description of what `failed=true` means in terms of functionality.